helpers/googleAnalytics.js

```diff
@@ -0,0 +1,116 @@
+/* eslint-disable no-console */
+const url = require('url');
+const crypto = require('crypto');
+const Request = require('request-promise');
+const ipaddr = require('ipaddr.js');
+const queryString = require('query-string');
+
+const SOCKET_POOL_SIZE = 32;
+const MAX_BATCH_SIZE = 10;
+const MAX_QUEUE_TIME = 60 * 60 * 1000;
+const FALLBACK_HOST = 'https://schul-cloud.org';
+const BATCH_HITS_ENDPOINT = 'https://www.google-analytics.com/batch';
+
+let queue = [];
+let flushTimer = [];
+const request = Request.defaults({
+  forever: true,
+  pool: {maxSockets: SOCKET_POOL_SIZE}
+});
+
+const emit = (hit) => {
+  queue.push(hit);
+  if(queue.length === MAX_BATCH_SIZE) {
+    flush();
+  } else {
+    // Schedules flush if no event is triggered within MAX_QUEUE_TIME
+    flushTimer = setTimeout(() => flush(), MAX_QUEUE_TIME);
+  }
+};
+
+const flush = () => {
+  if(flushTimer) {
+    clearTimeout(flushTimer);
+    flushTimer = null;
+  }
+  if(queue.length === 0) {
+    return;
+  }
+
+  let now = Date.now();
+  let body = queue.map(hit => {
+    // Adapts queue time parameter to the moment the hit is emitted
+    hit.qt = Math.round((now - hit.qt) / 1000);
+    return queryString.stringify(hit);
+  }).join('\r\n');
+  queue = [];
+
+  request({
+    method: 'POST',
+    uri: BATCH_HITS_ENDPOINT,
+    encoding: 'utf-8',
+    body,
+  }).catch(err => console.error(`Error while emitting Google Analytics hits (${err.message})`));
+};
+
+const anonymizeIp = (ip) => {
+  ip = ipaddr.parse(ip);
```

Comment:
can parsing fail under some circumstances?

Reply:
Only if the request IP is not valid, but I can also add error handling just in case.